### PR TITLE
Upload xcresults to LUCI cloud storage

### PR DESCRIPTION
**Pull request by cyanglaz**
Taking over from https://github.com/flutter/engine/pull/41644

fixes: https://github.com/flutter/flutter/issues/125823

Steps to verify
- [x] CI should pass with only the test script change, no xcresults are uploaded.
- [x] https://github.com/flutter/engine/pull/41652 should fail and xcresult should be uploaded.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by cyanglaz** [comment]
Scenario app test failure and result uploaded: https://github.com/flutter/engine/pull/41652

**Comment by cyanglaz** [comment]
Unittest failed and result uploaded: https://github.com/flutter/engine/pull/41929

File: testing/run_tests.py [hunk 1]
```diff
@@ -752,6 +753,12 @@ def run_objc_tests(ios_variant='ios_debug_sim_unopt', test_filter=None):
     ios_unit_test_dir = os.path.join(
         BUILDROOT_DIR, 'flutter', 'testing', 'ios', 'IosUnitTests'
     )
+
+    result_bundle_temp = tempfile.TemporaryDirectory(
```
Comment (gaaclarke) [inline testing/run_tests.py:757]: The way you use `TemporaryDirectory` is with a `with` statement to make sure that it gets cleaned up.

File: testing/run_tests.py [hunk 2]
```diff
@@ -760,13 +767,27 @@ def run_objc_tests(ios_variant='ios_debug_sim_unopt', test_filter=None):
         'xcodebuild '
         '-sdk iphonesimulator '
         '-scheme IosUnitTests '
-        "-destination name='" + new_simulator_name + "' "
+        '-resultBundlePath ' + result_bundle_path + " -destination name='" +
```
Comment (gaaclarke) [inline testing/run_tests.py:770]: Please follow the formatting of having a flag at the start of each line.

File: testing/run_tests.py [hunk 3]
```diff
@@ -760,13 +767,27 @@ def run_objc_tests(ios_variant='ios_debug_sim_unopt', test_filter=None):
...
       test_command[0] = test_command[0] + ' -only-testing:%s' % test_filter
     run_cmd(test_command, cwd=ios_unit_test_dir, shell=True)
+
+    # except:
+    # The LUCI environment may provide a variable containing a directory path
+    # for additional output files that will be uploaded to cloud storage.
+    # Upload the xcresult when the tests fail.
+    luci_test_outputs_path = os.environ.get('FLUTTER_TEST_OUTPUTS_DIR')
+    xcresult_bundle = os.path.join(result_bundle_temp, 'ios_embedding.xcresult')
+    print(xcresult_bundle)
```
Comment (gaaclarke) [inline testing/run_tests.py:785]: Stray print statement.

File: testing/scenario_app/run_ios_tests.sh [hunk 4]
```diff
@@ -41,25 +41,57 @@ fi
 # Can also be set via Simulator app Device > Rotate Device Automatically
 defaults write com.apple.iphonesimulator RotateWindowWhenSignaledByGuest -int 1
 
-cd $SRC_DIR/out/$FLUTTER_ENGINE/scenario_app/Scenarios
+SCENARIO_PATH=$SRC_DIR/out/$FLUTTER_ENGINE/scenario_app/Scenarios
+cd $SCENARIO_PATH
```
Comment (gaaclarke) [inline testing/scenario_app/run_ios_tests.sh:45]: I'd `pushd` and pop around here to make sure you don't effect any downstream commands.

File: testing/scenario_app/run_ios_tests.sh [hunk 5]
```diff
@@ -41,25 +41,57 @@ fi
...
+
+RESULT_BUNDLE_FOLDER="ios_scenario_xcresult"
+RESULT_BUNDLE_PATH="${SCENARIO_PATH}/${RESULT_BUNDLE_FOLDER}"
+
+# Zip and upload xcresult to luci.
+# First parameter ($1) is the zip output name.
+ZIP_AND_UPLOAD_XCRESULT_TO_LUCI () {
+  # Using absolute directory causes the zip containing all the sub directories.
```
Comment (gaaclarke) [inline testing/scenario_app/run_ios_tests.sh:53]: Can you clear this up please?  You are probably running into the difference between having a trailing slash or not fwiw.

File: testing/scenario_app/run_ios_tests.sh [hunk 6]
```diff
@@ -41,25 +41,57 @@ fi
...
+  # So use relative directory (./$RESULT_BUNDLE_FOLDER) instead.
+  echo $1
```
Comment (gaaclarke) [inline testing/scenario_app/run_ios_tests.sh:55]: Remove the echo or make it more useful by adding a label.

File: testing/scenario_app/run_ios_tests.sh [hunk 7]
```diff
@@ -41,25 +41,57 @@ fi
...
-cd $SRC_DIR/out/$FLUTTER_ENGINE/scenario_app/Scenarios
+SCENARIO_PATH=$SRC_DIR/out/$FLUTTER_ENGINE/scenario_app/Scenarios
+cd $SCENARIO_PATH
+
+RESULT_BUNDLE_FOLDER="ios_scenario_xcresult"
+RESULT_BUNDLE_PATH="${SCENARIO_PATH}/${RESULT_BUNDLE_FOLDER}"
+
+# Zip and upload xcresult to luci.
+# First parameter ($1) is the zip output name.
+ZIP_AND_UPLOAD_XCRESULT_TO_LUCI () {
```
Comment (gaaclarke) [inline testing/scenario_app/run_ios_tests.sh:52]: Make the function name lowercase please.

**Review by gaaclarke — CHANGES_REQUESTED** [review]

**Review by gaaclarke — APPROVED** [review]
lgtm, thanks!

File: testing/run_tests.py [hunk 8]
```diff
@@ -752,21 +753,43 @@ def run_objc_tests(ios_variant='ios_debug_sim_unopt', test_filter=None):
...
+          '-resultBundlePath ' + result_bundle_path + ' '
+          '-destination name=' + new_simulator_name + ' '
+          'test '
+          'FLUTTER_ENGINE=' + ios_variant
+      ]
+      if test_filter is not None:
+        test_command[0] = test_command[0] + ' -only-testing:%s' % test_filter
+      run_cmd(test_command, cwd=ios_unit_test_dir, shell=True)
+
+      # except:
```
Comment (cyanglaz) [inline testing/run_tests.py:778]: @gaaclarke Just found this is mistakenly commented out, I will create a patch to fix it.

Comment (cyanglaz) [reply testing/run_tests.py:778]: The PR landed before I can remove the auto_submit label :( I was a split second slower :(

Comment (gaaclarke) [reply testing/run_tests.py:778]: Ahh, nice job seeing the problem.  This wasn't showing up as an error in CI?  Do we need another test?

Comment (cyanglaz) [reply testing/run_tests.py:778]: It wouldn't as I also missed "try"

**Pull request merged**